src/utils/ScreenReader.ts

```diff
@@ -0,0 +1,281 @@
+   */
+  public async speakNode(node: Node, options?: SpeakOptions): Promise<void> {
+    const description = this.describe(node)
+    if (description) {
+      await this.speak(description, options)
+    }
+  }
+
+  /**
+   * Directly triggers speech of an event target. Resolves when speaking is done.
+   */
+  public async speakEventTarget(
+    target?: EventTarget,
+    { now = true }: SpeakOptions = {}
+  ): Promise<void> {
+    if (target && target instanceof Element) {
+      await this.speakNode(target, { now })
+    }
+  }
+
+  /**
+   * Generates a text string to be spoken for an element.
+   */
+  public describe(node: Node): string | undefined {
+    if (!(node instanceof Text) && !(node instanceof Element)) {
+      return
+    }
+
+    return this.cleanDescription(
+      node instanceof Text
+        ? this.describeText(node)
+        : this.describeElement(node)
+    )
+  }
+
+  private cleanDescription(
+    description: string | undefined
+  ): string | undefined {
+    if (!description) {
+      return
+    }
+    return description
+      .replace(/ {2}/g, ' ')
+      .replace(/^[. ]+/g, '')
+      .replace(/\. +\./g, '.')
+      .replace(/,\./g, '.')
+      .replace(/ +\./g, '.')
+      .replace(/ +,/g, ',')
+      .replace(/\.+/g, '.')
+      .replace(/ +$/g, '')
+      .replace(/^ +/g, '')
+  }
+
+  private describeText(node: Text): string | undefined {
+    return node.textContent || undefined
+  }
+
+  private describeElement(node: Element): string | undefined {
+    if (this.isHidden(node)) {
+      return
+    }
+
+    const terminator = this.isBlockElement(node) ? '.' : ''
+    const ariaLabel = node.getAttribute('aria-label')
+
+    if (ariaLabel) {
+      return ariaLabel + terminator
+    }
+
+    const ariaLabeledBy = node.getAttribute('aria-labeledby')
+
+    if (ariaLabeledBy) {
+      const element = document.getElementById(ariaLabeledBy)
+
+      if (element) {
+        const description = this.describe(element)
+
+        if (description) {
+          return description + terminator
+        }
+      }
+    }
+
+    return (
+      Array.from(node.childNodes)
+        .map(child => this.describe(child))
+        .filter(Boolean)
+        .join(' ') + terminator
+    )
+  }
+
+  /**
+   * Determines wheether `element` is a block or inline element.
+   */
+  private isBlockElement(element: Element): boolean {
+    return getComputedStyle(element).display === 'block'
+  }
+
+  /**
+   * Determines whether `element` is hidden from screen readers or not. Elements
+   * can be hidden either by setting the `aria-hidden` attribute or using CSS.
+   */
+  private isHidden(element: Element): boolean {
+    if (
+      element.hasAttribute('aria-hidden') &&
+      element.getAttribute('aria-hidden') !== 'false'
+    ) {
+      return true
+    }
+
+    const style = getComputedStyle(element)
+
+    if (style.display === 'none' || style.visibility === 'hidden') {
```

Comment:
It's possible there are more ways to hide an element than this, but these were the obvious ones.

Reply:
👍🏼 I promise not to hide elements using any other method. Eventually there may be elements animating from off-screen to on-screen, but the screen reader should be able to treat elements the same way as they do now.